### PR TITLE
[Bugfix][Feat] Add XML-ish grammar in EBNFComposer and fix misc bugs in Qwen3 detector

### PR DESCRIPTION
**Pull request by CatherineSue**
## Motivation
Follow up to #8260 

Here are some issues of `Qwen3XMLDetector`:

1. **EBNFComposer XML Support:** `tool_choice="required"` and specific tool selection fail because EBNFComposer only supported JSON/Pythonic formats, not XML
2. **Structural Tag Compatibility:** `strict=True` mode fails as Xgrammar doesn't support non-JSON formats with structural tags
  https://github.com/sgl-project/sglang/blob/12cb760a3773fe1a97d5a00fca26412f814f20fa/python/sglang/srt/function_call/function_call_parser.py#L157-L158
3. **Streaming Bug:** Incorrect tool call indexing in streaming mode
4. **Missing Tests:** No unit tests for Qwen3XMLDetector
5. **Incorrect Registry:** Should use `qwen3_coder` key since regular qwen3 models should use Qwen25Detector

This PR focuses on resolving item 1, 4 and 5. Item 2 is temporarily fixed with limitation.

## Modifications

**EBNFComposer XML Format Support**

- Added "xml" format alongside "pythonic" and "json"
- Created shared BASE_PRIMITIVE_GRAMMAR to eliminate duplication
- Simplified type mappings using base + overrides pattern:
  Before: 3 duplicate dictionaries
  After: DRY approach
```
  BASE_TYPE_MAPPING = {"string": "basic_string", ...}
  FORMAT_TYPE_OVERRIDES = {
      "pythonic": {"boolean": '"True" | "False"', "null": '"None"'},
      "xml": {"string": "xml_text"},
  }
```
- Refactored `_handle_enum()` and `_handle_type()` for better modularity

**Detector Refactoring**

- Renamed qwen3_detector.py → qwen3_coder_detector.py
- Renamed Qwen3XMLDetector → Qwen3CoderDetector
- Updated registry: "qwen3" → "qwen3_coder"
- Enhanced documentation to match codebase standards

**Structural Tag Support (Limited)**

- Added supports_structural_tag() method to BaseFormatDetector
- Override returns False for PythonicDetector and Qwen3CoderDetector

**Testing**

- Added `TestQwen3CoderDetector` with 11 test cases covering:
  - Basic detection and parsing
  - Streaming (simple, incomplete, incremental)
  - Type conversion and edge cases
  - EBNF generation and validation

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by CatherineSue** [comment]
UT result:
- override test_openai_function_calling to use exisitng qwen3-coder endpoint:
Screenshot 2025-07-25 at 11 00 11 AM

- override test_tool_choice to use exisiting qwen3-coder endpoint:
Screenshot 2025-07-25 at 10 59 42 AM

- EBNF and Qwen3CoderDetector
Screenshot 2025-07-25 at 11 17 22 AM

**Review by zhyncs — COMMENTED** [review]
QQ may we also handle this https://huggingface.co/Qwen/Qwen3-235B-A22B-Thinking-2507/blob/main/tokenizer_config.json#L229

File: python/sglang/srt/function_call/qwen3_coder_detector.py [hunk 1]
```diff
@@ -127,24 +126,26 @@ def _parse_block(self, block: str, tools: List[Tool]) -> List[ToolCallItem]:
...
         return EBNFComposer.build_ebnf(
             tools,
             individual_call_start_token=self.tool_call_start_token.replace("\n", "\\n"),
             individual_call_end_token=self.tool_call_end_token.replace("\n", "\\n"),
             tool_call_separator="\\n",
-            function_format="json",
+            function_format="xml",
+            call_rule_fmt='"<function={name}>\\n" {arguments_rule} "\\n</function>"',
+            key_value_rule_fmt='"<parameter={key}>\\n" {valrule} "\\n</parameter>"',
         )
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply python/sglang/srt/function_call/qwen3_coder_detector.py:151]: Want to be explicit here for more readability

**Comment by CatherineSue** [comment]
> QQ may we also handle this https://huggingface.co/Qwen/Qwen3-235B-A22B-Thinking-2507/blob/main/tokenizer_config.json#L229

I think this one can be handled with Qwen25Detector.

File: test/srt/test_function_call_parser.py [hunk 2]
```diff
@@ -1464,5 +1486,438 @@ def test_parse_streaming_multiple_tool_calls_with_multi_token_chunk(self):
...
+</function>
+</tool_call>"""
+
+        result = self.detector.detect_and_parse(model_output, tools=self.tools)
+        self.assertEqual(len(result.calls), 1)
+
+        params = json.loads(result.calls[0].parameters)
+        self.assertEqual(params["city"], "Dallas->TX")
+
+    def test_extract_tool_calls_fallback_no_tags(self):
```
Comment (CatherineSue) [inline test/srt/test_function_call_parser.py:1756]: NOTE: It seems the implementation in HF Qwen3-Coder model repo has a fall back if no tool_call tag is found. See: https://huggingface.co/Qwen/Qwen3-Coder-480B-A35B-Instruct-FP8/blob/main/qwen3coder_tool_parser.py#L229-L247

Our Qwen3CoderDetector currently doesn't support fallback so our test here is not consistent with the HF implementation. 

cc @yhyang201

**Comment by zhyncs** [comment]
In the model, the chat template will include <think> to force the model to think, but now qwen3's parser will distinguish whether the current content is written in reasoning or content based on whether the first token is <think>, which will cause the output of this model to have empty reasoning_content. Because this model is qwen 3 thinking, I am worried that using the qwen 2.5 parser may cause misunderstandings.

**Comment by CatherineSue** [comment]
I see your point. We should deal with the thinking part in the reasoning parser. 
For the function call parser, maybe we can update Qwen25Detector's registry to qwen to be more inclusive and less confusing.

**Comment by yhyang201** [comment]
Thank you very much for fixing the issue with the implementation.
